api/src/main/java/org/openmrs/module/appointmentscheduling/api/impl/AppointmentServiceImpl.java

```diff
@@ -1155,6 +1155,39 @@ public Appointment bookAppointment(Appointment appointment,
 				appointment);
 	}
 
+    @Override
+    public List<Appointment> getEarlyAppointments(Date fromDate,
+                                            Date toDate, Location location, Provider provider, AppointmentType appointmentType,
+                                            AppointmentStatus status, VisitType visitType) throws APIException {
+        List<Appointment> allCompletedAppointments = getAppointmentsByConstraints(fromDate,
+                toDate, location, provider, appointmentType, null, AppointmentStatus.COMPLETED);
+
+        List<Appointment> earlyAppointments = new ArrayList<Appointment>();
+        for (Appointment ap : allCompletedAppointments) {
+            if (ap.getVisit().getStartDatetime().before(ap.getTimeSlot().getEndDate())) {
+				earlyAppointments.add(ap);
+            }
+        }
+        return earlyAppointments;
+    }
+
+    @Override
+    public List<Appointment> getLateAppointments(Date fromDate,
+                                           Date toDate, Location location, Provider provider, AppointmentType appointmentType,
+                                           AppointmentStatus status, VisitType visitType) throws APIException {
+        List<Appointment> allCompletedAppointments = getAppointmentsByConstraints(fromDate,
+                toDate, location, provider, appointmentType, null, AppointmentStatus.COMPLETED);
+
+        List<Appointment> lateAppointments = new ArrayList<Appointment>();
+        for (Appointment ap : allCompletedAppointments) {
+            if (ap.getVisit().getStartDatetime().after(ap.getTimeSlot().getEndDate())) {
+				lateAppointments.add(ap);
```

Comment:
I'm a little confused here... this appears to determine if an appointment is late by checking if the the visit was after the appointment... this makes total sense to me, but then I don't think we need the new statuses for EARLY and LATE at all.